docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,123 @@
+- `Span<Span<char>>` &ndash; Nested `Span<T>` types would be of benefit in the parsing result of strings.
+
+## Runtime impact
+
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
+
+- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+- `stsfld` / `ldsfld` &ndash; Type fields of a ByRefLike parameter cannot be marked `static`.
+- `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; If this IL sequence resolves to a method implemented on `object` or default interface method, an error will occur during the attempt to box the instance.
+
+If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`. Sequences involving some of the above instructions are considered optimizations and represent cases that will remain valid regardless of a `T` being ByRefLike. See "Special IL Sequences" section below for details.
+
+The following instructions are already set up to support this feature since their behavior will fail as currently defined due to the inability to box a ByRefLike type.
+
+- `throw` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `unbox` / `unbox.any` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `isinst` &ndash; Will always place `null` on stack.
+- `castclass` &ndash; Will always throw `System.InvalidCastException`.
+
+## API Proposal
+
+Support for the following will be indicated by a new property. For .NET 7, the feature will be marked with `RequiresPreviewFeaturesAttribute` to indicate it is in [preview](https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md).
+
+```diff
+namespace System.Runtime.CompilerServices
+{
+    public static partial class RuntimeFeature
+    {
++        /// <summary>
++        /// Represents a runtime feature where byref-like types can be used in Generic parameters.
++        /// </summary>
++        public const string GenericsAcceptByRefLike = nameof(GenericsAcceptByRefLike);
+    }
+}
+```
+
+The compiler will need an indication for existing troublesome APIs where ByRefLike types will be permissable, but where the failure will be handled at runtime. An attribute will be created and added to these APIs.
+
+```csharp
+namespace System.Runtime.CompilerServices
+{
+    /// <summary>
+    /// Indicates to the compiler that constrain checks should be suppressed
+    /// and will instead be enforced at run-time.
+    /// </summary>
+    public sealed class SuppressConstraintChecksAttribute : Attribute
+    { }
+}
+```
+
+Troublesome APIs:
```

Comment:
I expect that there will be a lot more APIs like this that are troublesome. For example, `ArraySegment<T>` will need to allow T to be byref type (`static implicit operator ReadOnlySpan<T>(ArraySegment<T> segment)` would not compile otherwise) and APIs on `ArraySegment` that create arrays internally will be troublesome.

Reply:
> would not compile otherwise) and APIs on ArraySegment that create arrays internally will be troublesome.

I don't see why this is the case. The proposed `SuppressConstraintChecksAttribute` attribute would let the compiler know validation should be suppressed for this and handled at run-time. What am I missing with how this attribute will work?

Reply:
@jaredpar That is also how I understand this type too.  However, @jkotas mentioned the need for `ArraySegment<T>` to support ByRefLike types, which is the part I am not understanding. The `SuppressConstraintChecksAttribute` would make types that won't work in the context a run-time failure as opposed to the compiler handling the case. Or am I missing the distinction between `unsafe` and the flow of generic parameter constraints resolution?

Reply:
I assumed that the compiler will always enforce the constraints, and `SuppressConstraintChecksAttribute` will be only used to mark methods that may fail runtime.

If `SuppressConstraintChecksAttribute` were to suppress constraint enforcement, it opens a whole new set of failure modes. For example, inspecting a method signature using reflection may throw new kinds of exceptions since the signature may contain malformed types due to suppressed constraints.

Reply:
Specific example:
```
foreach (var m in typeof(Span<Span<int>>).GetMethods())
    foreach (var t in m.GetParameters())
        Console.WriteLine(t);
```
Does this throw an exception when it hits the operator that takes `ArraySegment`?

Reply:
I would not expect that to fail since it is being used for Reflection purposes. This is similar to what I was assuming was your perspective regarding `MakeGenericType` / `MakeGenericMethod` above.

Reply:
> I would not expect that to fail 

Reflection would need to return Type for `ArraySegment<Span<int>>` to make this succeed. Reflection only ever returns types with valid constraints. It means that `ArraySegment<T>` would need to allow T to be byref type.

Reply:
Good point. Another option would be to skip over the types that aren't valid. This approach would be less than ideal but given the scope of this API it might not be too bad. The tradeoff here is between special casing the Reflection API or adding the ByRefLike support to types that can never be instantiated. In our offline conversation, you mentioned `abstract` types fall into this as well so I think the latter approach makes the most sense right now.